Added Filtering Feature

HarshAggarwal1 opened the pull request:
<!-- If your PR fixes an open issue, use `Closes #999` to link your PR with the issue. #999 stands for the issue number you are fixing -->

## Fixes Issue

<!-- Please provide the issue number that you have resolved.  -->

<!-- Example: Fixes Issue #234 -->

## Changes proposed

<!-- List all the proposed changes in your PR -->

<!-- Mark all the applicable boxes. To mark the box as done follow the following conventions -->
<!--
[x] - Correct; marked as done
[X] - Correct; marked as done

[ ] - Not correct; marked as **not** done
-->

## Check List (Check all the applicable boxes) <!-- Follow the above conventions to check the box -->

- [x] My code follows the code style of this project.
- [x] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

## Screenshots

<!-- Add all the screenshots which support your changes -->

## Demo video


https://github.com/Taaveez/Taaveez-android/assets/88721766/a882e537-c063-4fdb-b5df-2fea12c613bc


## Note to reviewers

<!-- Add notes to reviewers if applicable -->

HarshAggarwal1 commented:
Hi @Suryansh1720001 , I have completed my work, I added dummy labels, you can edit them as per your need. By the way thanks for doing something in this project. Try this yourself also

Vaibhav2002 commented on app/src/main/java/com/itssuryansh/taaveez/Add_New_Content.kt:
```diff
@@ -81,12 +101,57 @@ class Add_New_Content : AppCompatActivity() {
             PoemDes?.undo()
         }
+        binding?.btnAddLabel?.setOnClickListener {
+            val dialogView = LayoutInflater.from(this).inflate(R.layout.diaglog_insert_label, null)
```
rather than instantiating the dialog, setting up the listeners, and else each and every time `btnAddLabel` is clicked, why don't you do the instantiation, setup and all only once, and on button press just display it and perform only the actions required specifically on button press.

Vaibhav2002 commented on app/src/main/java/com/itssuryansh/taaveez/Add_New_Content.kt:
```diff
@@ -38,6 +51,13 @@ class Add_New_Content : AppCompatActivity() {
         loadLocate()
         loadDayNight()
         super.onCreate(savedInstanceState)
+        selectedLabels = HashSet()
```
Can you try to create a database table for the labels, and fetch the labels from there?

Vaibhav2002 commented on app/src/main/java/com/itssuryansh/taaveez/Notes.kt:
```diff
@@ -60,20 +74,60 @@ class Notes : AppCompatActivity() {
             }
         }
+
+        //////////////////////////////////////////////////////////////////////////////////////////
```
What's this for?

HarshAggarwal1 replied:
Nothing, I was having problem to find my code scrolling so made this, I will remove it

Vaibhav2002 commented on app/src/main/java/com/itssuryansh/taaveez/Notes.kt:
```diff
@@ -60,20 +74,60 @@ class Notes : AppCompatActivity() {
-            NotesDao.fetchAllNotes().collect {
-                val list = ArrayList(it)
-                setupListOfDateINtoRecycleVIew(list, NotesDao)
+        val spinner = findViewById<Spinner>(R.id.spinnerFilter)
```
Why not create a separate function for this, and clean the `onCreate` a bit?

HarshAggarwal1 commented on app/src/main/java/com/itssuryansh/taaveez/Add_New_Content.kt:
```diff
@@ -81,12 +101,57 @@ class Add_New_Content : AppCompatActivity() {
+                .setNegativeButton("Done") { dialog, _ -> dialog.dismiss() }
+                .create()
+
+            val spinner = dialogView.findViewById<Spinner>(R.id.labelDropdown)
```
I think it does not cause any problem because both has same performance

HarshAggarwal1 replied:
Okay I will try to fix this, actually I code a lot in java, that's why I have a habit of it

HarshAggarwal1 replied:
I will see if I can change it

HarshAggarwal1 replied:
I am calling this via findViewByID in the AddContent function, and other codes are also done in same way
![Screenshot 2023-06-13 221701](https://github.com/Taaveez/Taaveez-android/assets/88721766/d697bc0d-18f0-4a00-9e12-3fdf9ffe0938)

Vaibhav2002 commented on app/src/main/java/com/itssuryansh/taaveez/NotesDatabase.kt:
```diff
@@ -5,7 +5,7 @@ import androidx.room.Database
 import androidx.room.RoomDatabase
 
-@Database(entities = [NotesEntity::class],version=2)
+@Database(entities = [NotesEntity::class],version=2, exportSchema = false)
```
As this app is on PlayStore, `exportSchema = false` can cause errors and crashes, which can affect user-experience. If its possible can you look for other ways, like Creating a Migration class?
Its also fine if you are unable to, will create another issue for this separately, when its merged

Vaibhav2002 commented on app/src/main/java/com/itssuryansh/taaveez/NotesEntity.kt:
```diff
@@ -11,5 +11,7 @@ data class NotesEntity(
     val Date:String="",
-    val CreatedDate:String=""
+    val CreatedDate:String="",
+    val Labels:String=""
```
Another table for Labels, and here should be the list of Ids of the labels

HarshAggarwal1 replied:
Bro actually I tried to make List here but during build it was showing some error, that's why I had to store it in string

HarshAggarwal1 replied:
I don't have screenshot of that, by the way you can also check, it will give some error

Vaibhav2002 replied:
I know about that error, you cannot store Lists in room

Vaibhav2002 replied:
Having a separate table for Labels and creating a relationship between Notes and Label Table seems a better fit for this use case

HarshAggarwal1 replied:
Bro can you make another issue for that, I don't have experience in databases, I have done other changes you said

Vaibhav2002 reviewed (changes requested):
Please make the required changes, as many as possible

Vaibhav2002 commented:
@HarshAggarwal1 just an advice, try to write commit messages which reflect the changes you've made

Vaibhav2002 commented on app/src/main/java/com/itssuryansh/taaveez/Edit_Content.kt:
```diff
@@ -46,14 +56,24 @@ class Edit_Content : AppCompatActivity() {
+        selectedLabels = HashSet()
+
+        // Label Suggestions
+        labelSuggestions = mutableListOf()
```
I know here you've done like this just for sample, would be better if they're saved in a separate Table, The note table would only have the id's of the labels and not the label as string

HarshAggarwal1 commented on app/src/main/java/com/itssuryansh/taaveez/Edit_Content.kt:
```diff
@@ -46,14 +56,24 @@ class Edit_Content : AppCompatActivity() {
         super.onCreate(savedInstanceState)
+
+        // Selected Labels
+        selectedLabels = HashSet()
```
Actually the problem is if anyone create labels by their own then there will be lots of labels , having predefined categories is better.

Vaibhav2002 replied:
Everyone will have their own set of labels, saved in their own device, it would cause any issues.

HarshAggarwal1 commented on app/src/main/java/com/itssuryansh/taaveez/Edit_Content.kt:
```diff
@@ -283,4 +312,74 @@ class Edit_Content : AppCompatActivity() {
     }
 
+    private fun updateLabels() {
+        val dialogView = LayoutInflater.from(this).inflate(R.layout.diaglog_insert_label, null)
```
It is necessary, let's say if you click outside dialog box, then all the labels will disappear whereas they are stored in selectedLabels, because in the logic we are not making Chip of the already existing label, that's why I think it is correct

Vaibhav2002 replied:
I meant only doing the creation of dialog, setup of listeners and all other things which need not be done each and every time, only once

HarshAggarwal1 commented on app/src/main/java/com/itssuryansh/taaveez/Notes.kt:
```diff
@@ -20,24 +20,38 @@ import androidx.annotation.RequiresApi
     @RequiresApi(Build.VERSION_CODES.JELLY_BEAN_MR1)
     override fun onCreate(savedInstanceState: Bundle?) {
 
+        labelSuggestions = mutableListOf()
```
Bro please make another issue for that

Vaibhav2002 commented:
> @HarshAggarwal1 just an advice, try to write commit messages which reflect the changes you've made

@HarshAggarwal1 please follow this

HarshAggarwal1 commented:
@Vaibhav2002 Bro I have made changes which I could do

Vaibhav2002 commented:
> @Vaibhav2002 Bro I have made changes which I could do

Then, please only mark those conversations as resolved which you have fixed, and leave the others as unresolved.

HarshAggarwal1 commented:
> > @Vaibhav2002 Bro I have made changes which I could do
> 
> Then, please only mark those conversations as resolved which you have fixed, and leave the others as unresolved.

Done

Vaibhav2002 commented:
> > > @Vaibhav2002 Bro I have made changes which I could do
> > 
> > 
> > Then, please only mark those conversations as resolved which you have fixed, and leave the others as unresolved.
> 
> Done

I think you should un resolve all the conversations, because you fixed none of them

HarshAggarwal1 commented:
> > > > @Vaibhav2002 Bro I have made changes which I could do
> > > 
> > > 
> > > Then, please only mark those conversations as resolved which you have fixed, and leave the others as unresolved.
> > 
> > 
> > Done
> 
> I think you should un resolve all the conversations, because you fixed none of them

Are you checking latest pushed code by me. I have

- Removed Unnecessary Logs
- Commented Code
- Removed unnecessary comments
- In edit content moved the filtering code to outer function from onCreate()
- findViewById() is done in other code also, if you want I will unresolve this

I just left database part. Bro you have written database issue in each comment, now how can I unesolve them 😂

Vaibhav2002 commented:
> > > > > @Vaibhav2002 Bro I have made changes which I could do
> > > > 
> > > > 
> > > > Then, please only mark those conversations as resolved which you have fixed, and leave the others as unresolved.
> > > 
> > > 
> > > Done
> > 
> > I think you should un resolve all the conversations, because you fixed none of them
> 
> Are you checking latest pushed code by me. I have
> 
> - Removed Unnecessary Logs
> - Commented Code
> - Removed unnecessary comments
> - In edit content moved the filtering code to outer function from onCreate()
> - findViewById() is done in other code also, if you want I will unresolve this
> 
> I just left database part. Bro you have written database issue in each comment, now how can I unesolve them 😂

Nevermind

HarshAggarwal1 commented:
> > > > > > @Vaibhav2002 Bro I have made changes which I could do
> > > > > 
> > > > > 
> > > > > Then, please only mark those conversations as resolved which you have fixed, and leave the others as unresolved.
> > > > 
> > > > 
> > > > Done
> > > 
> > > 
> > > I think you should un resolve all the conversations, because you fixed none of them
> > 
> > 
> > Are you checking latest pushed code by me. I have
> > 
> > * Removed Unnecessary Logs
> > * Commented Code
> > * Removed unnecessary comments
> > * In edit content moved the filtering code to outer function from onCreate()
> > * findViewById() is done in other code also, if you want I will unresolve this
> > 
> > I just left database part. Bro you have written database issue in each comment, now how can I unesolve them 😂
> 
> Nevermind You should atleast learn how to behave, If you cant learn how to code. This contribution will be accepted, please look into further contributions.

Thanks for letting me look for further contributions. I was just sharing my code status, my intention was not to offend you. By your comment to unresolve all, I thought you were looking at old code. But if you didn't like my comment, then I am sorry about that, and thanks for accepting this contribution. And again sorry, I was just trying to be friendly.

HarshAggarwal1 commented:
> > > > > > @Vaibhav2002 Bro I have made changes which I could do
> > > > > 
> > > > > 
> > > > > Then, please only mark those conversations as resolved which you have fixed, and leave the others as unresolved.
> > > > 
> > > > 
> > > > Done
> > > 
> > > 
> > > I think you should un resolve all the conversations, because you fixed none of them
> > 
> > 
> > Are you checking latest pushed code by me. I have
> > 
> > * Removed Unnecessary Logs
> > * Commented Code
> > * Removed unnecessary comments
> > * In edit content moved the filtering code to outer function from onCreate()
> > * findViewById() is done in other code also, if you want I will unresolve this
> > 
> > I just left database part. Bro you have written database issue in each comment, now how can I unesolve them 😂
> 
> Nevermind You should atleast learn how to behave, If you cant learn how to code. This contribution will be accepted, please look into further contributions.

Thanks for your input. But it would be great if you told me how and where I misbehaved, I think personal comments in professional in professional are not great. You told me to unresolved all of them in a very rude manner, that's why I listed all my changes and I was very friendly in all my comments. If you didn't like it then you must have said this cannot be accepted, but personal comments are not acceptable to me. May be you should also learn how to talk to other people.

If you think this work is non of use you can and you should not accept this contribution. I will report this on Hacktoberfest that where I am wrong.

HarshAggarwal1 commented:
No need for this shitty project
